File: 23 Merge k Sorted Lists/23 LL.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;
struct ListNode
{
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

/*
https://leetcode.com/problems/merge-k-sorted-lists/

Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
*/

class Solution
{
private:
	vector<ListNode*> data;
	bool pull_up(int pos = 1)
	{
		if (pos > data.size())
		{
			return false;
		}
		int child = pos << 1, i = pos;
		if (pull_up(child) && (data[i - 1]->val > data[child - 1]->val))
		{
			i = child;
		}
		if (pull_up(child + 1) && (data[i - 1]->val > data[child]->val))
		{
			i = child + 1;
		}
		if (i != pos)
		{
			swap(data[pos - 1], data[i - 1]);
			push_down(i);
		}
		return true;
	}
	void push_down(int pos = 1)
	{
		int child = pos << 1, i = pos;
		if ((child <= data.size()) && (data[i - 1]->val > data[child - 1]->val))
		{
			i = child;
		}
		if ((child < data.size()) && (data[i - 1]->val > data[child]->val))
		{
			i = child + 1;
		}
		if (i != pos)
		{
			swap(data[pos - 1], data[i - 1]);
			push_down(i);
		}
	}
public:
	ListNode* mergeKLists(vector<ListNode*>& lists)
	{
		ListNode result(0);
		vector<ListNode*>::iterator it = lists.begin();
		for (; it != lists.end(); ++it)
		{
			if (it[0] != NULL)
			{
				data.push_back(it[0]);
			}
		}
		Solution::pull_up();
		ListNode *slide = &result;
		while (data.size() != 0)
		{
			slide->next = data[0];
			slide = slide->next;
			if (slide->next != NULL)
			{
				data[0] = slide->next;
			}
			else
			{
				data[0] = data.back();
				data.pop_back();
			}
			Solution::push_down();
		}
		return result.next;
	}
};

/*
idea: using min heap tracing element
complexity: Time O(NlogN)
*/

int main(void)
{
	vector<ListNode*> lists;
	lists.push_back(new ListNode(3));
	lists.push_back(new ListNode(2));
	Solution engine;
	cout << engine.mergeKLists(lists)->val << '\n';
	return 0;
}

File: 23 Merge k Sorted Lists/23 LL.py
class ListNode(object):
	def __init__(self, x):
		self.val = x
		self.next = None

class Solution(object):
	minheap = list()
	def __pull_up(self, pos=1):
		if pos > len(self.minheap):
			return False
		flag = pos
		child = pos << 1
		if self.__pull_up(child) and self.minheap[flag - 1].val > self.minheap[child - 1].val:
			flag = child
		if self.__pull_up(child + 1) and self.minheap[flag - 1].val > self.minheap[child].val:
			flag = child + 1
		if flag != pos:
			self.minheap[flag - 1], self.minheap[pos - 1] = self.minheap[pos - 1], self.minheap[flag - 1]
			self.__push_down(flag)
		return True
	def __push_down(self, pos=1):
		flag = pos
		child = pos << 1
		if child <= len(self.minheap) and self.minheap[flag - 1].val > self.minheap[child - 1].val:
			flag = child
		if child < len(self.minheap) and self.minheap[flag - 1].val > self.minheap[child].val:
			flag = child + 1
		if flag != pos:
			self.minheap[flag - 1], self.minheap[pos - 1] = self.minheap[pos - 1], self.minheap[flag - 1]
			self.__push_down(flag)
	def mergeKLists(self, lists):
		for arr in lists:
			if arr is not None:
				self.minheap.append(arr)
		self.__pull_up()
		result = ListNode(0)
		slide = result
		while len(self.minheap) != 0:
			slide.next = self.minheap[0]
			slide = slide.next
			if slide.next:
				self.minheap[0] = slide.next
			else:
				self.minheap[0] = self.minheap[-1]
				self.minheap.pop()
			self.__push_down()
		return result.next

engine = Solution()
print engine.mergeKLists([ListNode(7), ListNode(5), ListNode(3)])